flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/RowSerializer.java

```diff
@@ -367,36 +367,42 @@ public int getVersion() {
 	/**
 	 * A {@link TypeSerializerSnapshot} for RowSerializer.
 	 */
-	// TODO not fully functional yet due to FLINK-17520
 	public static final class RowSerializerSnapshot extends CompositeTypeSerializerSnapshot<Row, RowSerializer> {
 
 		private static final int VERSION = 3;
 
-		private static final int VERSION_WITHOUT_ROW_KIND = 2;
+		private static final int LAST_VERSION_WITHOUT_ROW_KIND = 2;
 
-		private boolean legacyModeEnabled = false;
+		private int version = VERSION;
 
 		public RowSerializerSnapshot() {
 			super(RowSerializer.class);
 		}
 
 		RowSerializerSnapshot(RowSerializer serializerInstance) {
 			super(serializerInstance);
+			this.version = translateVersion(serializerInstance);
 		}
 
 		@Override
 		protected int getCurrentOuterSnapshotVersion() {
-			return VERSION;
+			return version;
 		}
 
 		@Override
 		protected void readOuterSnapshot(
 				int readOuterSnapshotVersion,
 				DataInputView in,
 				ClassLoader userCodeClassLoader) {
-			if (readOuterSnapshotVersion == VERSION_WITHOUT_ROW_KIND) {
-				legacyModeEnabled = true;
+			version = readOuterSnapshotVersion;
+		}
+
+		@Override
+		protected OuterSchemaCompatibility resolveOuterSchemaCompatibility(RowSerializer newSerializer) {
+			if (version == translateVersion(newSerializer)) {
```

Comment:
If I understood this correctly, this check doesn't need to be performed against the new serializer.

Since starting 1.11 the created new serializer is always writing in the new format (e.g. `legacyModeEnabled` is always `false`), combined with the fact that the version for the `RowSerializerSnapshot` was up-ticked in 1.11,
whether or not migration is needed is actually completely encoded / "piggy-backed" in the version of the `RowSerializerSnapshot`:

i.e. this should be able to be simplified to:
```suggestion
			if (readVersion <= LAST_VERSION_WITHOUT_ROW_KIND)) {
			    return OuterSchemaCompatibility.COMPATIBLE_AFTER_MIGRATION;
			}
			return OuterSchemaCompatibility.COMPATIBLE_AS_IS;
```